pkg/controller/signup_test.go

```diff
@@ -651,21 +651,71 @@ func (s *TestSignupSuite) TestVerifyCodeHandler() {
 			Key:   "code",
 			Value: "",
 		}
-		rr := initVerification(s.T(), handler, param, nil, userID, http.MethodGet, "/api/v1/signup/verification/")
+		rr := initVerification(s.T(), handler, param, nil, userID, "", http.MethodGet, "/api/v1/signup/verification/")
 
 		// Check the status code is what we expect.
 		require.Equal(s.T(), http.StatusBadRequest, rr.Code)
 	})
+
+	s.Run("usersignup stored by its username", func() {
+		// Create another UserSignup
+		otherUserID := uuid.Must(uuid.NewV4()).String()
+
+		otherUserSignup := &crtapi.UserSignup{
+			TypeMeta: v1.TypeMeta{},
+			ObjectMeta: v1.ObjectMeta{
+				Name:      "jsmith",
+				Namespace: configuration.Namespace(),
+				Annotations: map[string]string{
+					crtapi.UserSignupUserEmailAnnotationKey:        "jsmith@acme.com",
+					crtapi.UserVerificationAttemptsAnnotationKey:   "0",
+					crtapi.UserSignupVerificationCodeAnnotationKey: "999127",
+					crtapi.UserVerificationExpiryAnnotationKey:     time.Now().Add(10 * time.Second).Format(service.TimestampLayout),
+				},
+			},
+			Spec: crtapi.UserSignupSpec{
+				Userid:   otherUserID,
+				Username: "jsmith",
+			},
+			Status: crtapi.UserSignupStatus{},
+		}
+
+		err = s.FakeUserSignupClient.Tracker.Add(otherUserSignup)
+		require.NoError(s.T(), err)
+
+		// Create Signup controller instance.
+		ctrl := controller.NewSignup(s.Application)
+		handler := gin.HandlerFunc(ctrl.VerifyCodeHandler)
+
+		param := gin.Param{
+			Key:   "code",
+			Value: "999127",
+		}
+		rr := initVerification(s.T(), handler, param, nil, otherUserSignup.Spec.Userid, otherUserSignup.Spec.Username, http.MethodGet, "/api/v1/signup/verification")
```

Comment:
AFAICS, this test uses both valid `UserID` and `Username` args, so how do we know that the resource look-up was based on the latter and not the former?

Reply:
Good point, modified to just pass the username.